crates/typst-pdf/src/font.rs

```diff
@@ -160,9 +123,175 @@ pub(crate) fn write_fonts(ctx: &mut PdfContext) {
+                    font.advance(color_glyph.gid).unwrap_or(Em::new(0.0)).to_font_units();
+                widths.push(width);
+                // Create a fake page context for `write_frame`. We are only
+                // interested in the contents of the page.
+                let size = color_glyph.frame.size();
+                let mut page_ctx = PageContext::new(ctx, size);
+                page_ctx.bottom = size.y.to_f32();
+                page_ctx.content.start_color_glyph(width);
+                page_ctx.transform(
+                    // Make the Y axis go upwards, while preserving aspect ratio
+                    Transform::scale(Ratio::one(), -size.aspect_ratio())
+                        // Also move the origin to the top left corner
+                        .post_concat(Transform::translate(Abs::zero(), size.y)),
+                );
+                write_frame(&mut page_ctx, &color_glyph.frame);
+
+                // Retrieve the stream of the page and write it.
+                let stream = page_ctx.content.finish();
+                ctx.pdf.stream(instructions_stream_ref, &stream);
+
+                // Use this stream as instructions to draw the glyph.
+                glyphs_to_instructions.push(instructions_stream_ref);
+
+                gids.push(color_glyph.gid);
+            }
+
+            // Write the Type3 font object.
+            let mut pdf_font = ctx.pdf.type3_font(*subfont_id);
+            pdf_font.pair(Name(b"Resources"), ctx.type3_font_resources_ref);
+            pdf_font.bbox(color_font.bbox);
+            pdf_font.matrix([1.0 / scale_factor, 0.0, 0.0, 1.0 / scale_factor, 0.0, 0.0]);
+            pdf_font.first_char(0);
+            pdf_font.last_char((glyph_count - 1) as u8);
+            pdf_font.pair(Name(b"Widths"), widths_ref);
+            pdf_font.to_unicode(cmap_ref);
+            pdf_font.font_descriptor(descriptor_ref);
+
+            // Write the /CharProcs dictionary, that maps glyph names to
+            // drawing instructions.
+            let mut char_procs = pdf_font.char_procs();
+            for (gid, instructions_ref) in glyphs_to_instructions.iter().enumerate() {
+                char_procs
+                    .pair(Name(eco_format!("glyph{gid}").as_bytes()), *instructions_ref);
+            }
+            char_procs.finish();
+
+            // Write the /Encoding dictionary.
+            let names = (0..glyph_count)
+                .map(|gid| eco_format!("glyph{gid}"))
+                .collect::<Vec<_>>();
+            pdf_font
+                .encoding_custom()
+                .differences()
+                .consecutive(0, names.iter().map(|name| Name(name.as_bytes())));
+            pdf_font.finish();
+
+            // Encode a CMAP to make it possible to search or copy glyphs.
+            let glyph_set = ctx.glyph_sets.get_mut(&font).unwrap();
+            let mut cmap = UnicodeCmap::new(CMAP_NAME, SYSTEM_INFO);
+            for (index, gid) in gids.iter().enumerate() {
```

Comment:
can we just iterate over `color_font.glyphs[start..end]` here and skip the extra `gid` vector? (might want to store `&color_font.glyphs[start..end]` in a var)